### ISSUE TITLE
Inline HTML elements containing only whitespace are stripped from pasted content

### ISSUE THREAD

**Issue by gregsullivan** [WordPress/gutenberg#50898]
### Description

When pasting content into the block editor, elements containing only whitespace are removed. For example, the following pasted line:

``` html
The following space:<em> </em>will be <em>removed</em>.
```

is modified as follows on paste:

``` html
The following space:will be <em>removed</em>.
```

Stripping the tags makes sense to me, but I believe the whitespace should be retained.

I am encountering this on a magazine website where the content is pasted from Microsoft Word. The document is full of `<i>` elements containing only spaces, remnants of the editing process. These are removed on paste, merging words throughout the document.

### Step-by-step reproduction instructions

1. Go to the block editor.
2. Paste the following: `The following space:<em> </em>will be <em>removed</em>.`
3. Note that the space was removed.

### Environment info

- WordPress 6.2.2
- Gutenberg 15.8.1
- Twenty Twenty-Three 1.1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by gregsullivan** [comment]
You may need to paste the example sentence into a text editor and then into Gutenberg to ensure the HTML tags are parsed and not simply displayed.

**Comment by kathrynwp** [comment]
Hey there @gregsullivan - would it be feasible to clean up your Word documents to strip out the empty tags before pasting it into the editor?

**Comment by gregsullivan** [comment]
Hello @kathrynwp!

It would be feasible for me to clean up my client's Word documents, but not feasible for my client without adding meaningful time to the process of publishing articles on their website. They will be maintaining the site themselves beginning in July.

I don't expect this to be fixed by then, so I'm planning to write up instructions for how to paste from Microsoft Word to WordPress, covering:

* What is an empty HTML element?
* Saving from Word to an HTML document
* Using a text editor to replace-all within the document all instances of empty HTML elements (once per empty element type in cases where there is more than one kind)
* Copying from that HTML document and into WordPress

They are non-technical users, so this is asking a fair bit of them.

**Comment by ellatrix** [comment]
Hi! Could you share what is logged in the console when you paste the content? https://balsamiq.com/support/faqs/browserconsole/

**Comment by gregsullivan** [comment]
Hi Ella! Here you go:

From `use-paste-handler.js:103`:
```
Received HTML:

 
```

From `use-paste-handler.js:104`:
```
Received plain text:

 The following space:<em> </em>will be <em>removed</em>.
```

From `paste-handler.js:64`:
```
Processed inline HTML:

 The following space:will be <em>removed</em>.
```

My environment has updated to:

* WordPress 6.3.0
* Gutenberg 16.4.0

Thanks for looking at this, and for all your work on WordPress 6.3 (and beyond)!

**Comment by gusbova** [comment]
Hi, 

As of the latest version of WP 6.9.4, this problem still exists. If there is formatting on a blank space -- italics or bold for example -- Wordpress often just deletes that space entirely.

I run a news site and magazine, and this is such a hassle. The stories have already been edited and proofread repeatedly, and then typos are inserted at the last step. The formatting on spaces is invisible to us until we put it into Wordpress, and I can't just clear all the formatting in the documents because then I would lose intentional italics, hyperlinks, etc. 

We really need a fix on the Wordpress side here!

**Comment by gregsullivan** [comment]
@ellatrix @gusbova Inspired by @gusbova's comment, I decided to revisit this, and I've created a pull request that resolves the issue:

#76696

Thanks very much!

**Comment by gusbova** [comment]
Thank you!

**Comment by Mamaduka** [comment]
@gregsullivan, I also found a duplicate issue with a similar resolution - #74754. I think feedback from that PR would also be helpful here.

**Comment by gusbova** [comment]
Just want to chime in that I'm currently experiencing something odd where--at least in some of these cases--the problem still appears on the backend in the Wordpress visual editor. But, on the front end--in our case meaning the stories that the readers actually see--it is fixed. Almost like there was an incomplete patch?

**Comment by gregsullivan** [comment]
@gusbova Are you able to provide example text that triggers the above?

Just to confirm, your experience is:

- You paste in text
- The spaces are stripped
- You publish the post
- Spaces appear in the post
- When reloading the editor, spaces still do not appear in the editor

**Comment by gusbova** [comment]
Paste in text; space appears stripped; but after publishing--or when previewing it--the space appears. Reloading does not change it.

This should be two screenshots, one of the editor, one of the story as it looks published or previewed:

<img width="640" height="141" alt="Image" src="https://github.com/user-attachments/assets/bb26b6fd-87e8-427a-acb4-925e73250bc7" />

<img width="597" height="172" alt="Image" src="https://github.com/user-attachments/assets/369a8268-6836-47db-8895-041550fddbc8" />

Here's the text: https://docs.google.com/document/d/1074wQFH64vNmK9fHCcq4jmeb51TTfSfxgQerx6fzUIU/edit?usp=sharing

**Comment by gregsullivan** [comment]
Can you toggle "Edit as HTML" on the paragraph block with the selected text? When I paste it, I get this:

```html
<p>The right-of-entry letters reviewed by the<em> Observer</em> offer three options<br></p>
```

This feels like something different, related to font rendering for the italicized space versus the non-italicized one. Does your font stack differ between the editor and the front end?

**Comment by gusbova** [comment]
yes here it is pulled as html: `<p>The right-of-entry letters reviewed by the<em> Observer</em> offer three options` 

---

I can give another example. This is text that, when we published the story in March, definitely went live with these two spaces stripped on the front end. When I grab the text now, the two italicized spaces--in the first sentence and the second-to-last sentence--are still stripped in the back end when I paste it into a new post, but when I preview the front end it's fixed! 

`<p>The serpent in question was<em> </em>an adult western diamondback. Of the 10 species of rattler found in Texas, it is the largest—record holders can hit 7 feet, though 3 to 4 is standard—and by far the most common. It’s also the most famous. Diamondbacks play an outsize role in the Lone Star mythology: “Somehow talk about him seems bigger than elsewhere,” observed the folklorist J. Frank Dobie in <em>Rattlesnakes,</em> his 1965 compendium of state rattler lore.<em> </em>In many of these stories, informants recall rattlers as the menace of the rangeland, the creeping foe of the cowboy: a cold, unthinking set of venomous fangs, forever cocked, loaded, seeking the unwary ankle. “It is an unwritten law [among cowboys],” Dobie wrote, “that whoever sees a rattlesnake will kill it.”<br></p>`

---

I will try to figure out the font question.

**Comment by gregsullivan** [comment]
Just to double back, for this one:

```html
<p>The right-of-entry letters reviewed by the<em> Observer</em> offer three options<br></p>
```

That never should have been a problem, as there is other content inside the `em` element alongside the space. I should have been clearer about that in my last post.

For the more recent copy you pasted in above, that is a definite example of what triggers this issue. I tested that copy in WordPress 7.0 RC2 with and without the Gutenberg plugin activated. For me, the spaces were stripped in the same way as always.

**Comment by gusbova** [comment]
I sent your question about the font stack to the person who maintains our site, who understands this stuff much better than me.  I don't understand why I'm seeing what I'm seeing so I could be causing confusion.

He responded:

"He's asking if the fonts in the editor are different than the frontend. The answer is no, they are not, but more importantly I can recreate this issue on a fresh WP  6.9.4 and Twenty Twenty-Five theme. I've created a video for you / them to review:

https://www.loom.com/share/8b875ac505634485989bc52c50ff5e07"

**Comment by gregsullivan** [comment]
I created a fresh WP 6.9.4 install and used the default theme. I can't duplicate this: Spaces are stripped as before.

The video shows spaces not being stripped (as seen in code view, which shows the `<em> </em>` elements present). 6.9.4 still strips them. Are you using the Gutenberg plugin, patched or otherwise?

You could inspect the spaces in the editor (which are there, but not visible) to see whether CSS is being applied that hides them.

**Comment by gusbova** [comment]
Sorry for delay here, again from our website guy: 

"I was also testing with a brand new WP install 6.9.4 and NO additional plugins, so his experience doesn't seem to jive with mine. Attached is another screenshot showing that in visual block editor, the "<em> </em>" is actually stripped out in the rendered HTML. See attached.

here is the exact HTML code that is showing this behavior -


<p>This<em> </em>sentence<em> </em>should have spaces between each word.</p>


The above block shows this behavior on any site I've tested.

Again, no plugins, default theme, running locally. I attached site diagnostic info as well."

[local-diagnostic-formatting-issue.md](https://github.com/user-attachments/files/27730969/local-diagnostic-formatting-issue.md)

<img width="2231" height="725" alt="Image" src="https://github.com/user-attachments/assets/cf6db274-f4ea-4488-8857-7a6ee6bdea3f" />

**Comment by gregsullivan** [comment]
The screenshot shows the bug as I've always experienced it up to this point.

The unique part about the video is that switching to the code view showed `<em> </em>` elements. I have never experienced this.

I've tested in multiple browsers, and I can't duplicate what you're seeing. I'm not pasting the same copy you are, so it's possible there's something different about your copy: Maybe they're not standard spaces, etc.

But I can't duplicate this, and nothing has changed on my side (in WordPress 6.9.4 with no plugins activated, using the default theme).

**Comment by tyxla** [comment]
Also see #74740 for a near-identical issue reported more recently.
